### PR TITLE
Add migration file

### LINKED ISSUE #9: Support dumping and loading a set of caches from a "migration file"
**Brief Description of Feature:**
We should add rake tasks to:
- Dump the current set of caches that exist on ReadySet to a file that can be checked into version control
- Create caches on ReadySet based on this file

**Background:**
Our gem supports routing specific queries to ReadySet by annotating parts of an application's code. Because of how ReadySet works, it is not possible to create caches on the fly: cache creation for certain queries/tables can take quite a long time, and ReadySet is blocked from performing certain other actions while caches are being created.

As a result, cache creation must be a **separate step**, akin to database migrations. Because of the similarity to database migrations, we should provide a way for users to dump the current set of caches to a "migration file" of sorts, that will list all the caches that exist on ReadySet. The idea here is that the file can be checked into version control so the current set of desired caches is intentional and reproducible.

**Detailed Explanation:**
We should provide two separate rake tasks:
- `rails readyset:dump_current_caches`: Queries ReadySet for the current set of caches and dumps them to a file. This file should be stored in `db/readyset_migrate.rb`
- `rails readyset:migrate_caches`: Reads the `db/readyset_migrate.rb` file, drops any caches on ReadySet that don't exist in the file, and creates any caches from the file that don't already exist on ReadySet. This task should include logging so the user is aware of progress (I'm thinking we emit one log line per query, something like `Creating cache for <query>...` and then appending `done` at the end of the line when it's done.

---
## Optional information to add

**Proposed Solution:**
We should include a generator template to create the `readyset_migrate.rb` file. It would be nice to make this a ruby file and not just a yaml file because it is more extensible and allows us to do fancier things with it in the future (e.g. possibly allowing users to put ActiveRecord queries in the file instead of just SQL strings). Running `rails readyset:dump_current_caches` would use this template to create the file and load it with the desired info.

The rake tasks should use the correct database configuration based on `RAILS_ENV`, so users can set `RAILS_ENV` if they want to e.g. create caches on a remote ReadySet instance.

### PR DESCRIPTION
**Pull request by ethan-readyset**
This commit adds two rake tasks:
- `readyset:caches:dump`, which reads the caches that exist on the
  connected ReadySet instance and dumps them to a Ruby DSL file
- `readyset:caches:migrate`, which reads the aforementioned Ruby file,
  create the caches from the file that don't exist on ReadySet, and
  drops the caches on ReadySet that don't exist in the file

Closes #9

### CODE REVIEW

File: lib/readyset.rb [hunk 1]
```diff
@@ -88,7 +89,7 @@ def self.drop_cache!(name_or_id: nil, sql: nil)
   # @param sql_array [Array<Object>] the SQL array to be executed against ReadySet.
   # @return [PG::Result] the result of executing the SQL query.
   def self.raw_query(*sql_array) # :nodoc:
-    ActiveRecord::Base.connected_to(role: reading_role, shard: shard, prevent_writes: false) do
+    ActiveRecord::Base.connected_to(role: writing_role, shard: shard, prevent_writes: false) do
```
Comment (ethan-readyset) [inline lib/readyset.rb:92]: An unrelated bug that needed fixing. Our internal `raw_query` method needs to allow writes so we can invoke `CREATE CACHE` and `DROP CACHE` on ReadySet

File: lib/readyset/configuration.rb [hunk 2]
```diff
@@ -3,9 +3,10 @@
 
 module Readyset
   class Configuration
-    attr_accessor :shard
+    attr_accessor :migration_path, :shard
 
     def initialize
+      @migration_path = File.join(Rails.root, 'db/readyset_caches.rb')
```
Comment (ethan-readyset) [inline lib/readyset/configuration.rb:9]: Open to alternative suggestions for where to store this file!

File: spec/factories/proxied_query.rb [hunk 3]
```diff
@@ -6,11 +6,11 @@
     text { 'SELECT * FROM "t" WHERE ("x" = $1)' }
     supported { :yes }
 
-    factory :pending_query do
+    factory :pending_proxied_query do
```
Comment (ethan-readyset) [inline spec/factories/proxied_query.rb:9]: Another small, unrelated naming change

**Comment by ethan-readyset** [comment]
@helpotters Ready for some eyes! 👀 

I'd suggest testing this out end-to-end by doing something like this in the context of a rails application with our gem installed:
- Boot up ReadySet
- Create a few caches
- Run `rails readyset:caches:dump`
- Verify the cache migration file looks okay
- Drop one of the caches on ReadySet
- Remove one of the queries from the migration file (a different one than the cache dropped in the previous step)
- Run `rails readyset:caches:migrate`
- Verify the dropped cache was recreated and the cache removed from the migration file was dropped

**Review by jasobrown-rs — APPROVED** [review]
Ethan and I reviewed together, and I am +1
